### PR TITLE
[tune] refactor verbosity levels

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Adds a new verbosity level, so we now have four levels:

- `0` No reporting
- `1` Experiment-level updates
- `2` Brief trial-level updates
- `3` Detailed trial-level updates (like current default, stays the default).

Todo:

- [ ] Add output tests

This PR depends on #11748.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by krfricke** [comment]
I'll update this PR once #11749 is merged.

File: src/ray/raylet/raylet.cc [hunk 1]
```diff
@@ -69,17 +69,17 @@ Raylet::Raylet(boost::asio::io_service &main_service, const std::string &socket_
...
+          main_service, self_node_id_, object_manager_config, object_directory_,
+          [this](const ObjectID &object_id, const std::string &spilled_url,
+                 std::function<void(const ray::Status &)> callback) {
+            node_manager_.GetLocalObjectManager().AsyncRestoreSpilledObject(
+                object_id, spilled_url, callback);
+          },
+          [this](int64_t num_bytes_required) {
+            return node_manager_.GetLocalObjectManager().SpillObjectsOfSize(
+                num_bytes_required);
+          }),
```
Comment (richardliaw) [inline src/ray/raylet/raylet.cc:82]: remove :) ?I think you may want to update your clang (or uninstall)

Comment (krfricke) [reply src/ray/raylet/raylet.cc]: Argh. Updated clang-format and removed this from PR. I'll make sure to check this next time!

File: python/ray/tune/utils/log.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+from enum import Enum
+from typing import Any, Callable, Union
+
+
+class Verbosity(Enum):
+    MINIMAL = 0
+    EXPERIMENT = 1
+    TRIAL_NORM = 2
+    TRIAL_DETAILS = 3
```
Comment (richardliaw) [inline python/ray/tune/utils/log.py:9]: auto()?

File: python/ray/tune/tune.py [hunk 3]
```diff
@@ -188,8 +190,9 @@ def run(
         checkpoint_at_end (bool): Whether to checkpoint at the end of the
             experiment regardless of the checkpoint_freq. Default is False.
             This has no effect when using the Functional Training API.
-        verbose (int): 0, 1, or 2. Verbosity mode. 0 = silent,
-            1 = only status updates, 2 = status and trial results.
+        verbose (int): 0, 1, 2, or 3. Verbosity mode. 0 = silent,
+            1 = only status updates, 2 = status and brief trial results,
+            3 = status and detailed trial results.
```
Comment (richardliaw) [inline python/ray/tune/tune.py:195]: add a sentence about what the default is? "defaults to 3"

File: python/ray/tune/tune.py [hunk 4]
```diff
@@ -381,7 +385,8 @@ def run(
 
     if progress_reporter is None:
         if IS_NOTEBOOK:
-            progress_reporter = JupyterNotebookReporter(overwrite=verbose < 2)
+            progress_reporter = JupyterNotebookReporter(
+                overwrite=not has_verbosity(Verbosity.TRIAL_NORM))
```
Comment (richardliaw) [inline python/ray/tune/tune.py:389]: hmm I couldn't quite read what this means without looking at the underlying implementation.

File: python/ray/tune/progress_reporter.py [hunk 5]
```diff
@@ -621,3 +654,68 @@ def _get_trial_info(trial, parameters, metrics):
...
+        self.log_result(trial, result, error=False)
+
+    def on_trial_error(self, iteration: int, trials: List["Trial"],
+                       trial: "Trial", **info):
+        self.log_result(trial, trial.last_result, error=True)
+
+    def log_result(self, trial: "Trial", result: Dict, error: bool = False):
+        done = result.get("done", False) is True
+        last_print = self._last_print[trial]
+        if has_verbosity(Verbosity.TRIAL_DETAILS) and \
```
Comment (richardliaw) [inline python/ray/tune/progress_reporter.py:675]: hmm, this verbosity enum doesn't suggest any "ordering", as compared to say "very verbose, verbose, quiet". Any way to get that?

File: python/ray/tune/progress_reporter.py [hunk 6]
```diff
@@ -621,3 +654,68 @@ def _get_trial_info(trial, parameters, metrics):
...
+           (done or time.time() - last_print > DEBUG_PRINT_INTERVAL):
+            print("Result for {}:".format(self))
```
Comment (richardliaw) [inline python/ray/tune/progress_reporter.py:677]: Wouldn't this print a `Result for Progress Reporter callback` instead of referring to the trial?

File: python/ray/tune/progress_reporter.py [hunk 7]
```diff
@@ -621,3 +654,68 @@ def _get_trial_info(trial, parameters, metrics):
...
+            print("  {}".format(pretty_print(result).replace("\n", "\n  ")))
+            self._last_print[trial] = time.time()
+        elif has_verbosity(Verbosity.TRIAL_NORM) and (
+                done or time.time() - last_print > DEBUG_PRINT_INTERVAL):
+            info = ""
+            if done:
+                info = " This trial completed."
```
Comment (richardliaw) [inline python/ray/tune/progress_reporter.py:684]: This should reference the `str(trial)`

File: python/ray/tune/progress_reporter.py [hunk 8]
```diff
@@ -621,3 +654,68 @@ def _get_trial_info(trial, parameters, metrics):
...
+    def log_result(self, trial: "Trial", result: Dict, error: bool = False):
+        done = result.get("done", False) is True
+        last_print = self._last_print[trial]
+        if has_verbosity(Verbosity.TRIAL_DETAILS) and \
+           (done or time.time() - last_print > DEBUG_PRINT_INTERVAL):
+            print("Result for {}:".format(self))
+            print("  {}".format(pretty_print(result).replace("\n", "\n  ")))
+            self._last_print[trial] = time.time()
+        elif has_verbosity(Verbosity.TRIAL_NORM) and (
+                done or time.time() - last_print > DEBUG_PRINT_INTERVAL):
```
Comment (richardliaw) [inline python/ray/tune/progress_reporter.py:681]: if done or error or time.time() - last_print...?

File: python/ray/tune/progress_reporter.py [hunk 9]
```diff
@@ -621,3 +654,68 @@ def _get_trial_info(trial, parameters, metrics):
...
+                          "with parameters={trial.config}.{info}"
+
+            print(
+                message.format(
+                    trial=trial,
+                    error_file=error_file,
+                    metric_name=metric_name,
+                    metric_value=metric_value,
+                    result=print_result_str,
+                    info=info))
```
Comment (richardliaw) [inline python/ray/tune/progress_reporter.py:720]: maybe just use fstring here?

**Review by richardliaw — DISMISSED** [review]
(sorry for the late review!) Thanks for opening this PR; the callback is great. 

The only main concern for me is that the verbosity enum does not suggest any form of ordering (high verbosity or low verbosity); as a result, it's a bit hard to parse. Can we go with some other enumeration wording?

File: python/ray/tune/utils/log.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+def verbose_log(logger: Callable[[str], Any], level: Union[int, Verbosity],
+                message: str):
+    """Log `message` if specified level exceeds global verbosity level.
+
+    `logger` should be a Callable, e.g. `logger.info`. It can also be
+    `print` or a logger method of any other level - or any callable that
+    accepts a string.
+    """
+    if has_verbosity(level):
+        logger(message)
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/utils/log.py:36]: By passing the logger this should not happen, right? I.e. in `one.py` I instantiate a `logger`, which will log as `DEBUG one.py ...` - and if I call `verbose_log(logger.info, ...)` it will use the same logger and thus print the correct file. Right?

I thought about using custom loglevels for this but it seems to be much overhead and might interfere with other parts of Ray. Overall it seems introducing custom log levels is bad practice.

File: python/ray/tune/progress_reporter.py (hunk 6 above)
Comment (krfricke) [reply python/ray/tune/progress_reporter.py:677]: Right, thanks for spotting!

File: python/ray/tune/progress_reporter.py (hunk 7 above)
Comment (krfricke) [reply python/ray/tune/progress_reporter.py:684]: This string is appended to the message string, which already references the trial

File: python/ray/tune/progress_reporter.py (hunk 5 above)
Comment (krfricke) [reply python/ray/tune/progress_reporter.py:675]: Renamed to

```
class Verbosity(Enum):
    V0_MINIMAL = 0
    V1_EXPERIMENT = 1
    V2_TRIAL_NORM = 2
    V3_TRIAL_DETAILS = 3
```

(cannot start with integers). I think this will be nice for IDE auto completion. Users can still just use integers instead

File: python/ray/tune/utils/log.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/utils/log.py:9]: Since the int value is important (we want to be able to use `Verbosity.EXPERIMENT` and `int(1)` interchangeably, I'd rather specify them directly.

**Comment by krfricke** [comment]
Thanks for the comments!

**Review by richardliaw — COMMENTED** [review]
Looks good! Just need to add tests?

File: python/ray/tune/utils/log.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/utils/log.py:36]: Hmm, I think verbose_log will not use the same logger: 

```
2020-11-09 20:05:56,650	INFO log.py:36 -- Total run time: 2.39 seconds (2.35 seconds for the tuning loop).
```

[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/utils/log.py:36]: (which instead should show tune.py)

[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/utils/log.py:36]: I ran:  python examples/async_hyperband_example.py --smoke-test

**Comment by krfricke** [comment]
I added more thorough (end to end) output testing

File: python/ray/tune/utils/log.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/utils/log.py]: Sorry, this almost slipped. You're right, and that's because of our logger format `"%(asctime)s\t%(levelname)s %(filename)s:%(lineno)s -- %(message)s"` - I thought it was `name` instead of `filename`. 

Since we currently used it only in one location anyway I removed that utility function. It should be fine now.

File: python/ray/tune/tests/test_progress_reporter.py [hunk 11]
```diff
@@ -160,6 +160,48 @@ def f(config):
...
+Trial train_xxxxx_00002 reported acc=8 with parameters={'do': 'twice'}. """ + \
+    "This trial completed."
+
+VERBOSE_TRIAL_DETAIL = """+-------------------+----------+-------+----------+
+| Trial name        | status   | loc   | do       |
+|-------------------+----------+-------+----------|
+| train_xxxxx_00000 | RUNNING  |       | complete |
++-------------------+----------+-------+----------+"""
+
+VERBOSE_CMD = """from ray import tune
```
Comment (richardliaw) [inline python/ray/tune/tests/test_progress_reporter.py:180]: nit: I've been saving these as separate py files

**Review by richardliaw — APPROVED** [review]
